Sources/SourceKitLSP/Workspace.swift

```diff
@@ -103,14 +107,53 @@ public final class Workspace {
     capabilityRegistry: CapabilityRegistry,
     toolchainRegistry: ToolchainRegistry,
     buildSetup: BuildSetup,
+    workspaceType: WorkspaceType?,
     compilationDatabaseSearchPaths: [RelativePath],
     indexOptions: IndexOptions = IndexOptions(),
     reloadPackageStatusCallback: @escaping (ReloadPackageStatus) async -> Void
   ) async throws {
     var buildSystem: BuildSystem? = nil
     if let rootUrl = rootUri.fileURL, let rootPath = try? AbsolutePath(validating: rootUrl.path) {
-      if let buildServer = await BuildServerBuildSystem(projectRoot: rootPath, buildSetup: buildSetup) {
-        buildSystem = buildServer
+      if let workspaceType = workspaceType {
+        switch workspaceType {
+          case WorkspaceType.swiftpm:
+            if let swiftpm = await SwiftPMWorkspace(
+              url: rootUrl,
+              toolchainRegistry: toolchainRegistry,
+              buildSetup: buildSetup,
+              reloadPackageStatusCallback: reloadPackageStatusCallback
+            ) {
+              buildSystem = swiftpm
+            } else {
+              logger.error(
+                "Could not set up a swiftpm build system for workspace at '\(rootUri.forLogging)'"
+              )
+              buildSystem = nil
+            }
+          case WorkspaceType.buildserver:
+            if let buildServer = await BuildServerBuildSystem(projectRoot: rootPath, buildSetup: buildSetup) {
+              buildSystem = buildServer
+            } else {
+              logger.error(
+                "Could not set up a build server build system for workspace at '\(rootUri.forLogging)'"
+              )
+              buildSystem = nil
+            }
+          case WorkspaceType.compdb:
+            if let compdb = CompilationDatabaseBuildSystem(
+              projectRoot: rootPath,
+              searchPaths: compilationDatabaseSearchPaths
+            ) {
+              buildSystem = compdb
+            } else {
+              logger.error(
+                "Could not set up a compilation database build system for workspace at '\(rootUri.forLogging)'"
+              )
+              buildSystem = nil
+            }
+        }
+      } else if let buildServer = await BuildServerBuildSystem(projectRoot: rootPath, buildSetup: buildSetup) {
```

Comment:
If we can’t create a build server from the specified workspace type, I think we should still fall back to all the other selections. That way you can e.g. specify your default workspace type to be a compilation database in the VS Code settings but if you open a SwiftPM package, it would still work.